dali/operators/generic/one_hot.cuh

```diff
@@ -0,0 +1,47 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cstdint>
+
+namespace dali {
+
+namespace detail {
+
+struct SampleDesc {
+  uint64_t inner_vol, output_vol, inner_vol_classes;
+  void *out = nullptr;
+  const void *in = nullptr;
+};
+
+template <typename OutputType, typename InputType>
+__global__ void PopulateOneHot(OutputType on_value, OutputType off_value,
+                               const SampleDesc *samples) {
+  uint64_t out_index = static_cast<uint64_t>(blockIdx.x) * blockDim.x + threadIdx.x;
+  const auto &sample = samples[blockIdx.y];
+  if (out_index >= sample.output_vol) {
+    return;
+  }
+  auto *out = static_cast<OutputType*>(sample.out);
```

Comment:
Seems like this kernel is assuming that the grid size is enough to cover all the sample, which won't be true for big samples. You need to loop with a stride of the block size until you cover it all. Something like this:
```
const int64_t block_size = blockDim.x;
const int64_t grid_size = gridDim.x * block_size;
const int sample_idx = blockIdx.y;
const auto &sample = samples[sample_idx];
const int64_t offset = block_size * blockIdx.x;
const int64_t tid = threadIdx.x;
for (int64_t out_idx = offset + tid; out_idx < sample.size; out_idx += grid_size) {  // sample.size is the total volume of the sample
   // the rest of the code here
}
```

Reply:
Good point, it even works faster when inner_vol is huge (i.e. when axis is introduced at the beginning rather than at the end).